File: frontend/ios/Hacklahoma2026/TrailHealthKit.m
#import <React/RCTBridgeModule.h>
#import <HealthKit/HealthKit.h>

@interface TrailHealthKit : NSObject <RCTBridgeModule>
@property (nonatomic, strong) HKHealthStore *store;
@end

@implementation TrailHealthKit

RCT_EXPORT_MODULE();

- (instancetype)init {
  self = [super init];
  if (self) { _store = [[HKHealthStore alloc] init]; }
  return self;
}

+ (BOOL)requiresMainQueueSetup { return NO; }

#pragma mark - Type Helpers

- (HKQuantityType *)quantityTypeFor:(NSString *)typeId {
  static NSDictionary *map;
  static dispatch_once_t onceToken;
  dispatch_once(&onceToken, ^{
    map = @{
      @"steps": HKQuantityTypeIdentifierStepCount,
      @"active_calories": HKQuantityTypeIdentifierActiveEnergyBurned,
      @"exercise_minutes": HKQuantityTypeIdentifierAppleExerciseTime,
      @"distance_walk_run": HKQuantityTypeIdentifierDistanceWalkingRunning,
      @"flights_climbed": HKQuantityTypeIdentifierFlightsClimbed,
      @"resting_heart_rate": HKQuantityTypeIdentifierRestingHeartRate,
      @"hrv": HKQuantityTypeIdentifierHeartRateVariabilitySDNN,
      @"vo2max": HKQuantityTypeIdentifierVO2Max,
      @"water_intake": HKQuantityTypeIdentifierDietaryWater,
    };
  });
  NSString *identifier = map[typeId];
  return identifier ? [HKQuantityType quantityTypeForIdentifier:identifier] : nil;
}

- (HKCategoryType *)categoryTypeFor:(NSString *)typeId {
  static NSDictionary *map;
  static dispatch_once_t onceToken;
  dispatch_once(&onceToken, ^{
    map = @{
      @"sleep": HKCategoryTypeIdentifierSleepAnalysis,
      @"mindful_minutes": HKCategoryTypeIdentifierMindfulSession,
      @"stand_hours": HKCategoryTypeIdentifierAppleStandHour,
    };
  });
  NSString *identifier = map[typeId];
  return identifier ? [HKCategoryType categoryTypeForIdentifier:identifier] : nil;
}

- (HKUnit *)hkUnitFor:(NSString *)unit {
  if ([unit isEqualToString:@"count"]) return [HKUnit countUnit];
  if ([unit isEqualToString:@"kcal"]) return [HKUnit kilocalorieUnit];
  if ([unit isEqualToString:@"min"]) return [HKUnit minuteUnit];
  if ([unit isEqualToString:@"hr"]) return [HKUnit hourUnit];
  if ([unit isEqualToString:@"mi"]) return [HKUnit mileUnit];
  if ([unit isEqualToString:@"fl_oz"]) return [HKUnit fluidOunceUSUnit];
  if ([unit isEqualToString:@"bpm"]) return [[HKUnit countUnit] unitDividedByUnit:[HKUnit minuteUnit]];
  if ([unit isEqualToString:@"ms"]) return [HKUnit unitFromString:@"ms"];
  if ([unit isEqualToString:@"mL/kg/min"]) return [HKUnit unitFromString:@"ml/kg*min"];
  return [HKUnit countUnit];
}

- (NSDate *)parseISO:(NSString *)str {
  NSISO8601DateFormatter *f = [[NSISO8601DateFormatter alloc] init];
  f.formatOptions = NSISO8601DateFormatWithInternetDateTime | NSISO8601DateFormatWithFractionalSeconds;
  NSDate *d = [f dateFromString:str];
  if (!d) {
    f.formatOptions = NSISO8601DateFormatWithInternetDateTime;
    d = [f dateFromString:str];
  }
  return d ?: [NSDate date];
}

#pragma mark - Authorization

RCT_EXPORT_METHOD(requestAuthorization:(NSArray *)readTypes
                  resolver:(RCTPromiseResolveBlock)resolve
                  rejecter:(RCTPromiseRejectBlock)reject) {
  if (![HKHealthStore isHealthDataAvailable]) {
    reject(@"E_NO_HEALTH", @"HealthKit not available", nil);
    return;
  }
  NSMutableSet *types = [NSMutableSet set];
  for (NSString *t in readTypes) {
    HKQuantityType *qt = [self quantityTypeFor:t];
    if (qt) { [types addObject:qt]; continue; }
    HKCategoryType *ct = [self categoryTypeFor:t];
    if (ct) { [types addObject:ct]; continue; }
    if ([t isEqualToString:@"workout"]) { [types addObject:[HKWorkoutType workoutType]]; }
  }
  [self.store requestAuthorizationToShareTypes:nil readTypes:types completion:^(BOOL ok, NSError *err) {
    if (err) reject(@"E_AUTH", err.localizedDescription, err);
    else resolve(@(ok));
  }];
}

#pragma mark - Daily Sum

RCT_EXPORT_METHOD(queryDailySum:(NSString *)typeId date:(NSString *)date unit:(NSString *)unit
                  resolver:(RCTPromiseResolveBlock)resolve
                  rejecter:(RCTPromiseRejectBlock)reject) {
  HKQuantityType *qt = [self quantityTypeFor:typeId];
  if (!qt) { reject(@"E_TYPE", [NSString stringWithFormat:@"Unknown: %@", typeId], nil); return; }
  NSCalendar *cal = [NSCalendar currentCalendar];
  NSDate *d = [self parseISO:date];
  NSDate *start = [cal startOfDayForDate:d];
  NSDate *end = [cal dateByAddingUnit:NSCalendarUnitDay value:1 toDate:start options:0];
  NSPredicate *pred = [HKQuery predicateForSamplesWithStartDate:start endDate:end options:HKQueryOptionStrictStartDate];
  HKStatisticsQuery *q = [[HKStatisticsQuery alloc] initWithQuantityType:qt quantitySamplePredicate:pred options:HKStatisticsOptionCumulativeSum completionHandler:^(HKStatisticsQuery *query, HKStatistics *stats, NSError *err) {
    if (err) { reject(@"E_QUERY", err.localizedDescription, err); return; }
    double val = stats.sumQuantity ? [stats.sumQuantity doubleValueForUnit:[self hkUnitFor:unit]] : 0.0;
    resolve(@(val));
  }];
  [self.store executeQuery:q];
}

#pragma mark - Period Sum

RCT_EXPORT_METHOD(queryPeriodSum:(NSString *)typeId startDate:(NSString *)startDate endDate:(NSString *)endDate unit:(NSString *)unit
                  resolver:(RCTPromiseResolveBlock)resolve
                  rejecter:(RCTPromiseRejectBlock)reject) {
  HKQuantityType *qt = [self quantityTypeFor:typeId];
  if (!qt) { reject(@"E_TYPE", [NSString stringWithFormat:@"Unknown: %@", typeId], nil); return; }
  NSDate *start = [self parseISO:startDate];
  NSDate *end = [self parseISO:endDate];
  NSPredicate *pred = [HKQuery predicateForSamplesWithStartDate:start endDate:end options:HKQueryOptionStrictStartDate];
  HKStatisticsQuery *q = [[HKStatisticsQuery alloc] initWithQuantityType:qt quantitySamplePredicate:pred options:HKStatisticsOptionCumulativeSum completionHandler:^(HKStatisticsQuery *query, HKStatistics *stats, NSError *err) {
    if (err) { reject(@"E_QUERY", err.localizedDescription, err); return; }
    double val = stats.sumQuantity ? [stats.sumQuantity doubleValueForUnit:[self hkUnitFor:unit]] : 0.0;
    resolve(@(val));
  }];
  [self.store executeQuery:q];
}

#pragma mark - Latest Value

RCT_EXPORT_METHOD(queryLatestValue:(NSString *)typeId unit:(NSString *)unit
                  resolver:(RCTPromiseResolveBlock)resolve
                  rejecter:(RCTPromiseRejectBlock)reject) {
  HKQuantityType *qt = [self quantityTypeFor:typeId];
  if (!qt) { reject(@"E_TYPE", [NSString stringWithFormat:@"Unknown: %@", typeId], nil); return; }
  NSSortDescriptor *sort = [NSSortDescriptor sortDescriptorWithKey:HKSampleSortIdentifierStartDate ascending:NO];
  HKSampleQuery *q = [[HKSampleQuery alloc] initWithSampleType:qt predicate:nil limit:1 sortDescriptors:@[sort] resultsHandler:^(HKSampleQuery *query, NSArray *samples, NSError *err) {
    if (err) { reject(@"E_QUERY", err.localizedDescription, err); return; }
    if (samples.count == 0) { resolve([NSNull null]); return; }
    HKQuantitySample *s = samples.firstObject;
    resolve(@([s.quantity doubleValueForUnit:[self hkUnitFor:unit]]));
  }];
  [self.store executeQuery:q];
}

#pragma mark - Workout Count

RCT_EXPORT_METHOD(queryWorkoutCount:(NSString *)startDate endDate:(NSString *)endDate
                  resolver:(RCTPromiseResolveBlock)resolve
                  rejecter:(RCTPromiseRejectBlock)reject) {
  NSDate *start = [self parseISO:startDate];
  NSDate *end = [self parseISO:endDate];
  NSPredicate *pred = [HKQuery predicateForSamplesWithStartDate:start endDate:end options:HKQueryOptionStrictStartDate];
  HKSampleQuery *q = [[HKSampleQuery alloc] initWithSampleType:[HKWorkoutType workoutType] predicate:pred limit:HKObjectQueryNoLimit sortDescriptors:nil resultsHandler:^(HKSampleQuery *query, NSArray *samples, NSError *err) {
    if (err) { reject(@"E_QUERY", err.localizedDescription, err); return; }
    resolve(@(samples.count));
  }];
  [self.store executeQuery:q];
}

#pragma mark - Sleep

RCT_EXPORT_METHOD(querySleepDuration:(NSString *)date
                  resolver:(RCTPromiseResolveBlock)resolve
                  rejecter:(RCTPromiseRejectBlock)reject) {
  HKCategoryType *ct = [HKCategoryType categoryTypeForIdentifier:HKCategoryTypeIdentifierSleepAnalysis];
  NSCalendar *cal = [NSCalendar currentCalendar];
  NSDate *d = [self parseISO:date];
  NSDate *noon = [cal dateBySettingHour:12 minute:0 second:0 ofDate:d options:0];
  NSDate *prevEvening = [cal dateByAddingUnit:NSCalendarUnitHour value:-18 toDate:noon options:0];
  NSPredicate *pred = [HKQuery predicateForSamplesWithStartDate:prevEvening endDate:noon options:HKQueryOptionStrictStartDate];
  NSSortDescriptor *sort = [NSSortDescriptor sortDescriptorWithKey:HKSampleSortIdentifierStartDate ascending:YES];
  HKSampleQuery *q = [[HKSampleQuery alloc] initWithSampleType:ct predicate:pred limit:HKObjectQueryNoLimit sortDescriptors:@[sort] resultsHandler:^(HKSampleQuery *query, NSArray *samples, NSError *err) {
    if (err) { reject(@"E_QUERY", err.localizedDescription, err); return; }
    double total = 0;
    for (HKCategorySample *s in samples) {
      if (s.value != HKCategoryValueSleepAnalysisInBed) {
        total += [s.endDate timeIntervalSinceDate:s.startDate];
      }
    }
    resolve(@(total / 3600.0));
  }];
  [self.store executeQuery:q];
}

#pragma mark - Stand Hours

RCT_EXPORT_METHOD(queryStandHours:(NSString *)date
                  resolver:(RCTPromiseResolveBlock)resolve
                  rejecter:(RCTPromiseRejectBlock)reject) {
  HKCategoryType *ct = [HKCategoryType categoryTypeForIdentifier:HKCategoryTypeIdentifierAppleStandHour];
  NSCalendar *cal = [NSCalendar currentCalendar];
  NSDate *d = [self parseISO:date];
  NSDate *start = [cal startOfDayForDate:d];
  NSDate *end = [cal dateByAddingUnit:NSCalendarUnitDay value:1 toDate:start options:0];
  NSPredicate *pred = [HKQuery predicateForSamplesWithStartDate:start endDate:end options:HKQueryOptionStrictStartDate];
  HKSampleQuery *q = [[HKSampleQuery alloc] initWithSampleType:ct predicate:pred limit:HKObjectQueryNoLimit sortDescriptors:nil resultsHandler:^(HKSampleQuery *query, NSArray *samples, NSError *err) {
    if (err) { reject(@"E_QUERY", err.localizedDescription, err); return; }
    NSInteger stood = 0;
    for (HKCategorySample *s in samples) {
      if (s.value == HKCategoryValueAppleStandHourStood) stood++;
    }
    resolve(@(stood));
  }];
  [self.store executeQuery:q];
}

#pragma mark - Activity Summary

RCT_EXPORT_METHOD(queryActivitySummary:(NSString *)date
                  resolver:(RCTPromiseResolveBlock)resolve
                  rejecter:(RCTPromiseRejectBlock)reject) {
  NSCalendar *cal = [NSCalendar currentCalendar];
  NSDate *d = [self parseISO:date];
  NSDateComponents *comp = [cal components:(NSCalendarUnitYear | NSCalendarUnitMonth | NSCalendarUnitDay) fromDate:d];
  comp.calendar = cal;
  NSPredicate *pred = [HKQuery predicateForActivitySummaryWithDateComponents:comp];
  HKActivitySummaryQuery *q = [[HKActivitySummaryQuery alloc] initWithPredicate:pred resultsHandler:^(HKActivitySummaryQuery *query, NSArray<HKActivitySummary *> *summaries, NSError *err) {
    if (err) { reject(@"E_QUERY", err.localizedDescription, err); return; }
    if (summaries.count == 0) { resolve([NSNull null]); return; }
    HKActivitySummary *s = summaries.firstObject;
    resolve(@{
      @"activeCalories": @([s.activeEnergyBurned doubleValueForUnit:[HKUnit kilocalorieUnit]]),
      @"exerciseMinutes": @([s.appleExerciseTime doubleValueForUnit:[HKUnit minuteUnit]]),
      @"standHours": @([s.appleStandHours doubleValueForUnit:[HKUnit countUnit]]),
      @"moveGoal": @([s.activeEnergyBurnedGoal doubleValueForUnit:[HKUnit kilocalorieUnit]]),
      @"exerciseGoal": @([s.appleExerciseTimeGoal doubleValueForUnit:[HKUnit minuteUnit]]),
      @"standGoal": @([s.appleStandHoursGoal doubleValueForUnit:[HKUnit countUnit]]),
    });
  }];
  [self.store executeQuery:q];
}

@end

File: frontend/ios/Hacklahoma2026/TrailHealthKit.swift
import Foundation
import HealthKit
import React

@objc(TrailHealthKit)
class TrailHealthKit: NSObject {
  
  private let store = HKHealthStore()
  
  @objc static func requiresMainQueueSetup() -> Bool { return false }
  
  @objc func requestAuthorization(
    _ readTypes: [String],
    resolver resolve: @escaping RCTPromiseResolveBlock,
    rejecter reject: @escaping RCTPromiseRejectBlock
  ) {
    guard HKHealthStore.isHealthDataAvailable() else {
      reject("E_NO_HEALTH", "HealthKit not available", nil)
      return
    }
    var types = Set<HKObjectType>()
    for t in readTypes {
      if let qt = quantityType(for: t) { types.insert(qt) }
      else if let ct = categoryType(for: t) { types.insert(ct) }
      else if t == "workout" { types.insert(HKObjectType.workoutType()) }
    }
    store.requestAuthorization(toShare: nil, read: types) { ok, err in
      if let err = err { reject("E_AUTH", err.localizedDescription, err) }
      else { resolve(ok) }
    }
  }
  
  @objc func queryDailySum(
    _ typeId: String, date: String, unit: String,
    resolver resolve: @escaping RCTPromiseResolveBlock,
    rejecter reject: @escaping RCTPromiseRejectBlock
  ) {
    guard let qt = quantityType(for: typeId) else { reject("E_TYPE", "Unknown: \(typeId)", nil); return }
    let cal = Calendar.current
    let d = iso(date) ?? Date()
    let start = cal.startOfDay(for: d)
    let end = cal.date(byAdding: .day, value: 1, to: start)!
    let pred = HKQuery.predicateForSamples(withStart: start, end: end, options: .strictStartDate)
    let q = HKStatisticsQuery(quantityType: qt, quantitySamplePredicate: pred, options: .cumulativeSum) { _, stats, err in
      if let err = err { reject("E_QUERY", err.localizedDescription, err); return }
      resolve(stats?.sumQuantity()?.doubleValue(for: self.hkUnit(for: unit)) ?? 0.0)
    }
    store.execute(q)
  }
  
  @objc func queryPeriodSum(
    _ typeId: String, startDate: String, endDate: String, unit: String,
    resolver resolve: @escaping RCTPromiseResolveBlock,
    rejecter reject: @escaping RCTPromiseRejectBlock
  ) {
    guard let qt = quantityType(for: typeId) else { reject("E_TYPE", "Unknown: \(typeId)", nil); return }
    let start = iso(startDate) ?? Date()
    let end = iso(endDate) ?? Date()
    let pred = HKQuery.predicateForSamples(withStart: start, end: end, options: .strictStartDate)
    let q = HKStatisticsQuery(quantityType: qt, quantitySamplePredicate: pred, options: .cumulativeSum) { _, stats, err in
      if let err = err { reject("E_QUERY", err.localizedDescription, err); return }
      resolve(stats?.sumQuantity()?.doubleValue(for: self.hkUnit(for: unit)) ?? 0.0)
    }
    store.execute(q)
  }
  
  @objc func queryLatestValue(
    _ typeId: String, unit: String,
    resolver resolve: @escaping RCTPromiseResolveBlock,
    rejecter reject: @escaping RCTPromiseRejectBlock
  ) {
    guard let qt = quantityType(for: typeId) else { reject("E_TYPE", "Unknown: \(typeId)", nil); return }
    let sort = NSSortDescriptor(key: HKSampleSortIdentifierStartDate, ascending: false)
    let q = HKSampleQuery(sampleType: qt, predicate: nil, limit: 1, sortDescriptors: [sort]) { _, samples, err in
      if let err = err { reject("E_QUERY", err.localizedDescription, err); return }
      guard let s = samples?.first as? HKQuantitySample else { resolve(nil); return }
      resolve(s.quantity.doubleValue(for: self.hkUnit(for: unit)))
    }
    store.execute(q)
  }
  
  @objc func queryWorkoutCount(
    _ startDate: String, endDate: String,
    resolver resolve: @escaping RCTPromiseResolveBlock,
    rejecter reject: @escaping RCTPromiseRejectBlock
  ) {
    let start = iso(startDate) ?? Date()
    let end = iso(endDate) ?? Date()
    let pred = HKQuery.predicateForSamples(withStart: start, end: end, options: .strictStartDate)
    let q = HKSampleQuery(sampleType: HKObjectType.workoutType(), predicate: pred, limit: HKObjectQueryNoLimit, sortDescriptors: nil) { _, samples, err in
      if let err = err { reject("E_QUERY", err.localizedDescription, err); return }
      resolve(samples?.count ?? 0)
    }
    store.execute(q)
  }
  
  @objc func querySleepDuration(
    _ date: String,
    resolver resolve: @escaping RCTPromiseResolveBlock,
    rejecter reject: @escaping RCTPromiseRejectBlock
  ) {
    guard let ct = HKObjectType.categoryType(forIdentifier: .sleepAnalysis) else { reject("E_TYPE", "Sleep N/A", nil); return }
    let cal = Calendar.current
    let d = iso(date) ?? Date()
    let noon = cal.date(bySettingHour: 12, minute: 0, second: 0, of: d)!
    let prevEvening = cal.date(byAdding: .hour, value: -18, to: noon)!
    let pred = HKQuery.predicateForSamples(withStart: prevEvening, end: noon, options: .strictStartDate)
    let sort = NSSortDescriptor(key: HKSampleSortIdentifierStartDate, ascending: true)
    let q = HKSampleQuery(sampleType: ct, predicate: pred, limit: HKObjectQueryNoLimit, sortDescriptors: [sort]) { _, samples, err in
      if let err = err { reject("E_QUERY", err.localizedDescription, err); return }
      var total: Double = 0
      for s in (samples as? [HKCategorySample]) ?? [] {
        if s.value != HKCategoryValueSleepAnalysis.inBed.rawValue {
          total += s.endDate.timeIntervalSince(s.startDate)
        }
      }
      resolve(total / 3600.0)
    }
    store.execute(q)
  }
  
  @objc func queryStandHours(
    _ date: String,
    resolver resolve: @escaping RCTPromiseResolveBlock,
    rejecter reject: @escaping RCTPromiseRejectBlock
  ) {
    guard let ct = HKObjectType.categoryType(forIdentifier: .appleStandHour) else { reject("E_TYPE", "Stand N/A", nil); return }
    let cal = Calendar.current
    let d = iso(date) ?? Date()
    let start = cal.startOfDay(for: d)
    let end = cal.date(byAdding: .day, value: 1, to: start)!
    let pred = HKQuery.predicateForSamples(withStart: start, end: end, options: .strictStartDate)
    let q = HKSampleQuery(sampleType: ct, predicate: pred, limit: HKObjectQueryNoLimit, sortDescriptors: nil) { _, samples, err in
      if let err = err { reject("E_QUERY", err.localizedDescription, err); return }
      let stood = (samples as? [HKCategorySample])?.filter { $0.value == HKCategoryValueAppleStandHour.stood.rawValue }.count ?? 0
      resolve(stood)
    }
    store.execute(q)
  }
  
  @objc func queryActivitySummary(
    _ date: String,
    resolver resolve: @escaping RCTPromiseResolveBlock,
    rejecter reject: @escaping RCTPromiseRejectBlock
  ) {
    let cal = Calendar.current
    let d = iso(date) ?? Date()
    let components = cal.dateComponents([.year, .month, .day], from: d)
    let pred = HKQuery.predicateForActivitySummary(with: components)
    let q = HKActivitySummaryQuery(predicate: pred) { _, summaries, err in
      if let err = err { reject("E_QUERY", err.localizedDescription, err); return }
      guard let s = summaries?.first else { resolve(nil); return }
      resolve([
        "activeCalories": s.activeEnergyBurned.doubleValue(for: .kilocalorie()),
        "exerciseMinutes": s.appleExerciseTime.doubleValue(for: .minute()),
        "standHours": s.appleStandHours.doubleValue(for: .count()),
        "moveGoal": s.activeEnergyBurnedGoal.doubleValue(for: .kilocalorie()),
        "exerciseGoal": s.appleExerciseTimeGoal.doubleValue(for: .minute()),
        "standGoal": s.appleStandHoursGoal.doubleValue(for: .count()),
      ])
    }
    store.execute(q)
  }
  
  // MARK: - Helpers
  private func quantityType(for id: String) -> HKQuantityType? {
    let map: [String: HKQuantityTypeIdentifier] = [
      "steps": .stepCount, "active_calories": .activeEnergyBurned,
      "exercise_minutes": .appleExerciseTime, "distance_walk_run": .distanceWalkingRunning,
      "flights_climbed": .flightsClimbed, "resting_heart_rate": .restingHeartRate,
      "hrv": .heartRateVariabilitySDNN, "vo2max": .vo2Max, "water_intake": .dietaryWater,
    ]
    guard let id = map[id] else { return nil }
    return HKQuantityType.quantityType(forIdentifier: id)
  }
  private func categoryType(for id: String) -> HKCategoryType? {
    let map: [String: HKCategoryTypeIdentifier] = [
      "sleep": .sleepAnalysis, "mindful_minutes": .mindfulSession, "stand_hours": .appleStandHour,
    ]
    guard let id = map[id] else { return nil }
    return HKCategoryType.categoryType(forIdentifier: id)
  }
  private func hkUnit(for unit: String) -> HKUnit {
    switch unit {
    case "count": return .count(); case "kcal": return .kilocalorie()
    case "min": return .minute(); case "hr": return .hour()
    case "mi": return .mile(); case "fl_oz": return .fluidOunceUS()
    case "bpm": return HKUnit(from: "count/min")
    case "ms": return .secondUnit(with: .milli)
    case "mL/kg/min": return HKUnit(from: "ml/kg*min")
    default: return .count()
    }
  }
  private func iso(_ s: String) -> Date? {
    let f = ISO8601DateFormatter()
    f.formatOptions = [.withInternetDateTime, .withFractionalSeconds]
    return f.date(from: s) ?? ISO8601DateFormatter().date(from: s)
  }
}
